pkg/applications/tbtc/chain/ethereum/ethereum.go

```diff
@@ -0,0 +1,182 @@
+package ethereum
```

Comment:
I was going back and forward on this one in my head for 3 days and unfortunately, I don't think it is a good move and the original idea of keeping it all in `pkg/chain/ethereum` was better IMO.

The fact we need to cast the value returned from `CreateContract` to a specific implementation is the first warning signal. Looking close, we see that `CreateContract` exposes private fields of the base `ethereum` package hoping that the extension will treat them in a way that `ethereum` package expects. There is no difference between this and making those fields public, and we know they should not be public. The extension may not follow the rules and e.g. use a mutex in a way that will block all other calls. That's another warning signal.

Thinking more about it, `go.mod` is importing `tbtc` contract bindings anyway, so I think it's fine to expect - for now - that the client exposes etheruem API for those bindings and the responsibility of an extension is to implement app-specific actions in the correct way, using those bindings. If we decide in the future, we really need those applications to be pluggable, we should refactor base `ethereum` package to allow plugging-in new chain extension without having the base `ethereum` package exposing private properties.

What I would suggest for now - let's do it in a way that is not perfect, but safer and good enough, and not blocking us for future if we decide we really need chain extensions to be pluggable. My proposition - that is actually the same or very similar to what we had here originally:
- Let's keep it logic in a separate file, but in the base ethereum package, e.g. `pkg/chain/ethereum/tbtc.go`.
- Let's have a separate `TBTC` interface with TBTC-specific functions exposed by `Handle` in `pkg/chain/chain.go`, similar to `BlockCounter` or `StakeMonitor`
- The logic of the extension - the code using the chain and defining how to react to certain events - should stay separate, like it's done here now.

Reply:
Ok, agree but I think there is no need to:

> Let's have a separate TBTC interface with TBTC-specific functions exposed by Handle in pkg/chain/chain.go, similar to BlockCounter or StakeMonitor

I think we should rather keep the TBTC interface at the point of use, in the extension file, as we are doing it now. This way we follow the philosophy of Go interfaces and make the code more flexible. Also we don't need to expose the TBTC interface as part of our regular `Handle` interface because the TBTC part is used exclusively by the extension.

Reply:
I may not see all the parts yet, but if you say it's going to work I like it!